File: ex6-matrix-max-openmp.c
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#define VALIDATE 0

int max(const size_t n, const int * restrict);
int validate_mm(const size_t n, const int * restrict, const int);
void usage(char**);

int main(int argc, char **argv)
{
    int *A,max_val;
    size_t i,j,n;
    double t0,t1;

    if(argc==2)
        sscanf(argv[1],"%zu",&n);
    else {
        usage(argv);
        return 1;
    }

    A = (int*)malloc(n*n*sizeof(int));
    for(i=0; i<n; ++i)
        for(j=0; j<n; ++j) A[i*n+j]=i*n+j;

    t0 = omp_get_wtime();
    max_val = max(n,A);
    t1 = omp_get_wtime();

#if VALIDATE
    if(!validate_mm(n,A,max_val)) {
        printf("Validation failed.\n");
        return 1;
    }
#endif

    printf("Total time taken: %f.\n",t1-t0);

    free(A);
    return 0;
}

int max(const size_t n, const int * restrict A)
{
    size_t i,j;
    int max_val=A[0];
    #pragma omp parallel for default(none) shared(max_val,A) private(i,j)
    for(i=0; i<n; ++i)
        for(j=0; j<n; ++j)
            if(A[i*n+j]>max_val) {
                #pragma omp critical
                {
                    max_val=A[i*n+j];
                }
            }
    return max_val;
}

int validate_mm(const size_t n, const int * restrict A, const int max_val)
{
    size_t i,j;
    int tmax=A[0];
    for(i=0; i<n; ++i)
        for(j=0; j<n; ++j)
            if(A[i*n+j]>tmax) tmax=A[i*n+j];
    return (max_val!=tmax) ? 0 : 1;
}

void usage(char **argv)
{
    printf("Usage: %s <length>\n",argv[0]);
}

File: f90/ex6-matrix-max-openmp.f90
program main
    use omp_lib
    implicit none
    integer, dimension (:, :), allocatable :: A
    integer :: i, j, n, max_val
    real :: u
    real(8) :: t0, t1
    character(32) :: arg

    if(command_argument_count() .eq. 1) then
        call get_command_argument(1, arg)
        read (arg, *) n
    else
        call get_command_argument(0, arg)
        call usage(arg)
        stop 1
    endif

    call random_seed()

    allocate(A(n, n))
    do i=1,n
        do j=1,n
            call random_number(u)
            A(i, j) = floor(u*huge(n))
        end do
    end do

    t0 = omp_get_wtime()
    ! max()/maxval() are intrinsic functions
    max_val = kmax(A)
    t1 = omp_get_wtime()

    print *, "max(A) = ", max_val
    print *, "Total time taken: ", t1-t0

    deallocate(A)

contains
    integer function kmax(A)
        integer, dimension(:, :), intent (in) :: A
        integer :: i, j, n, m
        n = size(A, 1)
        m = size(A, 2)
        kmax = A(1, 1)
        !$omp parallel do default(none) shared(max_val,n,m,A) private(i,j)
        do i=1,n
            do j=1,m
                !$omp critical
                if(A(i, j) .gt. kmax) then
                    kmax = A(i, j)
                endif
                !$omp end critical
            end do
        end do
        !$omp end parallel do
    end function

end program

subroutine usage(arg)
    character(32) :: arg
    print *, "Usage: ", arg(1:len_trim(arg)), " <length>"
end subroutine
